sentry-core/src/main/java/io/sentry/core/CircularFifoQueue.java

```diff
@@ -0,0 +1,402 @@
+package io.sentry.core;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.AbstractCollection;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+
+/**
+ * CircularFifoQueue is a first-in first-out queue with a fixed size that replaces its oldest
+ * element if full.
+ *
+ * <p>The removal order of a {@link CircularFifoQueue} is based on the insertion order; elements are
+ * removed in the same order in which they were added. The iteration order is the same as the
+ * removal order.
+ *
+ * <p>The {@link #add(Object)}, {@link #remove()}, {@link #peek()}, {@link #poll}, {@link
+ * #offer(Object)} operations all perform in constant time. All other operations perform in linear
+ * time or worse.
+ *
+ * <p>This queue prevents null objects from being added.
+ *
+ * @since 4.0
+ * @version $Id$
+ */
+class CircularFifoQueue<E> extends AbstractCollection<E> implements Queue<E>, Serializable {
```

Comment:
I took this from `sentry-java`.
It's been working fine there for years but it had no tests.
We could add some tests but I'd merge this first because right now an app would run out of memory if using this since breadcrumbs is uncapped

Reply:
The old SDK has every method acessing this as `symchronized` though so we need to rethink this.

Reply:
you could use `ReentrantLock/ReentrantReadWriteLock`

Reply:
fixed it using `SynchronizedQueue/SynchronizedCollection` from apache